operator/controllers/btpoperator_controller.go

```diff
@@ -347,3 +417,283 @@ func (r *BtpOperatorReconciler) watchBtpOperatorUpdatePredicate() predicate.Func
+				return err, true
+			}
+		}
+
+		return nil, len(list.Items) > 0
+	}
+
+	for _, namespace := range namespaces.Items {
+		err, instancesLeft := anyLeft(namespace.Name, instanceGvk)
+		if err != nil {
+			return err, true
+		}
+		err, bindingsLeft := anyLeft(namespace.Name, bindingGvk)
+		if err != nil {
+			return err, true
+		}
+		if instancesLeft == false && bindingsLeft == false {
+			return nil, false
+		}
+	}
+
+	return nil, true
+}
+
+func (r *BtpOperatorReconciler) handleSoftDelete(ctx context.Context, namespaces *corev1.NamespaceList) error {
+	logger := log.FromContext(ctx)
+	logger.Info("hard delete failed. trying to perform soft delete")
+
+	if err := r.softDelete(ctx, &instanceGvk); err != nil {
+		logger.Error(err, "soft deletion of instances failed")
+		return err
+	}
+
+	if err := r.softDelete(ctx, &bindingGvk); err != nil {
+		logger.Error(err, "hard deletion of bindings failed")
+		return err
+	}
+
+	if err := r.cleanUpAllBtpOperatorResources(ctx, namespaces); err != nil {
+		logger.Error(err, "failed to remove related installed resources")
+		return err
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) softDelete(ctx context.Context, gvk *schema.GroupVersionKind) error {
+	listGvk := *gvk
+	listGvk.Kind = gvk.Kind + "List"
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(listGvk)
+
+	if err := r.List(ctx, list); err != nil {
+		err = fmt.Errorf("%w; could not list in soft delete", err)
+		return err
+	}
+
+	isBinding := gvk.Kind == btpOperatorServiceBinding
+	for _, item := range list.Items {
+		item.SetFinalizers([]string{})
+		if err := r.Update(ctx, &item); err != nil {
+			return err
+		}
+
+		if isBinding {
+			secret := &corev1.Secret{}
+			secret.Name = item.GetName()
+			secret.Namespace = item.GetNamespace()
+			if err := r.Delete(ctx, secret); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := r.List(ctx, list); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		if len(list.Items) > 0 {
+			return fmt.Errorf("finalizers deletion assurance has failed")
+		}
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) handlePreDelete(ctx context.Context) error {
+	deployment := &v1.Deployment{}
+	deployment.Namespace = chartNamespace
+	deployment.Name = deploymentName
+	if err := r.Delete(ctx, deployment); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	mutatingWebhook := &admissionregistrationv1.MutatingWebhookConfiguration{}
+	if err := r.DeleteAllOf(ctx, mutatingWebhook, labelFilter); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	validatingWebhook := &admissionregistrationv1.ValidatingWebhookConfiguration{}
+	if err := r.DeleteAllOf(ctx, validatingWebhook, labelFilter); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) cleanUpAllBtpOperatorResources(ctx context.Context, namespaces *corev1.NamespaceList) error {
```

Comment:
collecting all GVKs from the cluster just to delete a few connected to the BTP Service Operator chart seems like overdo. Could GVKs be limited only to these present in the chart?

Reply:
BTP Manager holds a local copy of a chart and we could fetch all the required GVKs from there. Let's discuss it f2f.